processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionExecutor.java

```diff
@@ -140,25 +139,72 @@ public CarbonCompactionExecutor(Map<String, TaskBlockInfo> segmentMapping,
-                  sourceSegProperties, destinationSegProperties, false));
+        // during update there may be a chance that the cardinality may change within the segment
+        // which may lead to failure while converting the row, so get all the blocks present in a
+        // task and then split into multiple lists of same key length and create separate
+        // RawResultIterator for each list of same key length. If all the blocks have same keylength
+        // , then make a single RawResultIterator for all the blocks
+        List<List<TableBlockInfo>> listOfTableBlocksBasedOnKeyLength =
+            getListOfTableBlocksBasedOnKeyLength(list);
+        for (List<TableBlockInfo> tableBlockInfoList : listOfTableBlocksBasedOnKeyLength) {
+          Collections.sort(tableBlockInfoList);
+          LOGGER.info(
+              "for task -" + task + "- in segment id -" + segmentId + "- block size is -" + list
+                  .size());
+          queryModel.setTableBlockInfos(tableBlockInfoList);
+          if (sortingRequired) {
+            resultList.get(CarbonCompactionUtil.UNSORTED_IDX).add(new RawResultIterator(
+                executeBlockList(tableBlockInfoList, segmentId, task, configuration),
+                getSourceSegmentProperties(
+                    Collections.singletonList(tableBlockInfoList.get(0).getDataFileFooter())),
+                destinationSegProperties, false));
+          } else {
+            resultList.get(CarbonCompactionUtil.SORTED_IDX).add(new RawResultIterator(
+                executeBlockList(tableBlockInfoList, segmentId, task, configuration),
+                getSourceSegmentProperties(
+                    Collections.singletonList(tableBlockInfoList.get(0).getDataFileFooter())),
+                destinationSegProperties, false));
+          }
         }
       }
     }
     return resultList;
   }
 
+  /**
+   * This method returns the List of TableBlockInfoList, where each listOfTableBlockInfos will have
+   * same keySize
+   * @param tableBlockInfos List of tableBlock Infos
+   * @return
+   */
+  private List<List<TableBlockInfo>> getListOfTableBlocksBasedOnKeyLength(
+      List<TableBlockInfo> tableBlockInfos) {
+    List<List<TableBlockInfo>> listOfTableBlockInfoListOnKeySize = new ArrayList<>();
+    // create a map of keySizeInBytes got from dataFileFooter
+    Map<Integer, List<TableBlockInfo>> keySizeToTableBlockInfoMap = new HashMap<>();
+    // iterate through tableBlockInfos
+    for (TableBlockInfo tableBlock : tableBlockInfos) {
+      // get the keySizeInBytes for the dataFileFooter
+      int keySizeInBytes =
+          getSourceSegmentProperties(Collections.singletonList(tableBlock.getDataFileFooter()))
+              .getDimensionKeyGenerator().getKeySizeInBytes();
+      // get the value for the key keySizeInBytes
+      List<TableBlockInfo> tempBlockInfoList = keySizeToTableBlockInfoMap.get(keySizeInBytes);
```

Comment:
better to use keySizeToTableBlockInfoMap.contains

Reply:
anyway if i use contains also, i need to create a list inside and put to map, it is one and  the same, so i think it is ok